### ISSUE TITLE
Using `transform` with zeroshot

### ISSUE THREAD

**Issue by timdhu** [MaartenGr/BERTopic#1797]
Hi Maarten. Thanks for all the work on this brilliant library.

I think I’ve identified a bug in the zeroshot feature. If I follow [the tutorial](https://maartengr.github.io/BERTopic/getting_started/zeroshot/zeroshot.html) and compare the results to a `transform`:

```
from datasets import load_dataset
from pandas import DataFrame

from bertopic import BERTopic
from bertopic.representation import KeyBERTInspired

# We select a subsample of 5000 abstracts from ArXiv
dataset = load_dataset("CShorten/ML-ArXiv-Papers")["train"]
docs = dataset["abstract"][:5_000]

# We define a number of topics that we know are in the documents
zeroshot_topic_list = ["Clustering", "Topic Modeling", "Large Language Models"]

# We fit our model using the zero-shot topics
# and we define a minimum similarity. For each document,
# if the similarity does not exceed that value, it will be used
# for clustering instead.
topic_model = BERTopic(
    embedding_model="thenlper/gte-small", 
    min_topic_size=15,
    zeroshot_topic_list=zeroshot_topic_list,
    zeroshot_min_similarity=.85,
    representation_model=KeyBERTInspired()
)
topics, probabilities = topic_model.fit_transform(docs)

# Calculate topics and probabilities using the transform method
new_topics, new_probabilities = topic_model.transform(docs)

result = DataFrame(data={"documents": docs, "topics": topics, "probabilities": probabilities, "new_topics": new_topics, "new_probabilities": new_probabilities})
```

When I look at `result.head(10)` I see different results:

![Screenshot 2024-02-08 at 10 44 49](https://github.com/MaartenGr/BERTopic/assets/58509816/196343d5-b98f-4ac5-9ba1-bab49a92d335)

From what I understand the results shouldn’t be identical, because `fit_transform` uses the topics defined by the original models (UMAP, HDBSCAN, etc.), while `transform` uses cosine similarity with topic embeddings. However, a difference of 2 seems weirdly common. If I look at the average difference between the two predictions `(result["topics"] - result["new_topics"]).value_counts()`, most of the time the difference is -2:

```
-2     3114
 0      353
-9       77
-25      58
 2       55
       ... 
 31       1
-50       1
 26       1
 19       1
 52       1
Name: count, Length: 102, dtype: int64
```

I think that this is because `topic_model.topic_embeddings_` are incorrect: they seem to be off by two positions. Perhaps this is related to [this issue](https://github.com/MaartenGr/BERTopic/issues/1771).

As a workaround I’m reidentifying the outlier topic and recalculating topic embeddings:

```
from bertopic._bertopic import TopicMapper

embeddings = topic_model.embedding_model.embed(docs)

topic_info = topic_model.get_topic_info()
outlier_label = topic_info.loc[topic_info["Name"].str.startswith("-1"), "Topic"].values[0]

# Restructure topic labels, sizes, representations and representative documents
for attribute_name in ["topic_labels_", "topic_sizes_", "topic_representations_", "representative_docs_"]:
    attribute = getattr(topic_model, attribute_name)
    attribute[-1] = attribute.pop(outlier_label)
    attribute = {key if key < outlier_label else key - 1: value for key, value in attribute.items()}
    attribute = dict(sorted(attribute.items()))
    setattr(topic_model, attribute_name, attribute)

topics = [topic if topic < outlier_label else -1 if topic == outlier_label else topic - 1 for topic in topics]

# Recreate topic embeddings
topic_model.topic_embeddings_ = None
topic_model.topic_mapper_ = TopicMapper(list(topic_model.topic_labels_.keys()))  # The topic mapper is used to map topics between merged models, we override it here to avoid errors
topic_model._create_topic_vectors(documents=DataFrame(data={"sentences": docs, "Topic": topics, "ID": range(len(docs))}), embeddings=embeddings)

topic_model._outliers = 1  # We have outliers in the model
```

The results look sensible:

```
new_topics, new_probabilities = topic_model.transform(docs)
result = DataFrame(data={"documents": docs, "topics": topics, "probabilities": probabilities, "new_topics": new_topics, "new_probabilities": new_probabilities})
```

`result.head(10)`:

![Screenshot 2024-02-08 at 10 41 44](https://github.com/MaartenGr/BERTopic/assets/58509816/1872365c-7e9c-4237-8115-1b209e0a6957)

`(result["topics"] - result["new_topics"]).value_counts()`:

```
 0     3470
-10      75
-24      59
-11      57
-26      55
       ... 
 47       1
 17       1
 14       1
 27       1
 54       1
Name: count, Length: 107, dtype: int64
```

**Comment by MaartenGr** [comment]
Thanks for sharing this extensive write-up! It is indeed related to https://github.com/MaartenGr/BERTopic/issues/1771 and the same issue. The outlier cluster (-1) is expected to be in the 0th position throughout BERTopic. However, with zero-shot topic modeling, it generally ends up in the >0th position instead since zero-shot topics are set first. 

I believe the best (and easiest!) approach here would simply be to swap the order of the zeroshot topics with those of the regular topic model. 

I think but I am not entirely sure to switching this:

https://github.com/MaartenGr/BERTopic/blob/da52e2c40d56c5351dce3c4af26978d10db3b913/bertopic/_bertopic.py#L3627

to this:

```python
        merged_model = BERTopic.merge_models([self, zeroshot_model], min_similarity=1)
```


would solve the issue. If you, or anyone else, has time to test this that would be great! Otherwise, I'll put it on my backlog which can take a while since I do not work on BERTopic full-time but merely weekends/late hours.

**Comment by timdhu** [comment]
Thanks for the speedy reply :) I've mentioned in the other issue that I think there is a separate bug for merging models. 

In terms of the outlier thing, I think swapping the two models around in the call would solve it, but the nice thing about putting the `zeroshot_model` first is that those topics/labels are prioritised over cluster topics/labels... Is there instead a way  to add an "outlier" cluster to the zeroshot model, which is just all the documents that haven't been assigned a zeroshot topic?

Perhaps the zeroshot model here

https://github.com/MaartenGr/BERTopic/blob/da52e2c40d56c5351dce3c4af26978d10db3b913/bertopic/_bertopic.py#L3591

can be fitted with all docs, where docs without topics are given a value of -1, or the `BaseCluster` class can be switched out for an alternative that assigned unknown documents as -1. What do you think?

**Comment by MaartenGr** [comment]
Hmmm, I think the easiest way to fix this issue is then to simply move the outlier class back to position -1 whilst keeping the non-outlier topics in the same order. I want to keep the functionality of the zero-shot model the same such that also non-zero-shot topics are found. 

That way, we keep the core functionality whilst only needing to move the outlier topic back to position -1. If we were to aggregate all documents not assigned to a zero-shot topic to the outliers, then that would remove significant information for users wanting to further develop those topics.

**Comment by timdhu** [comment]
Ok that makes sense. I’m afraid I can’t see the best way to implement this, is it something that should be the responsibility of `merge_models`?

**Comment by MaartenGr** [comment]
> Ok that makes sense. I’m afraid I can’t see the best way to implement this, is it something that should be the responsibility of merge_models?

I created a PR in https://github.com/MaartenGr/BERTopic/pull/1804 that should solve both issues, the ordering of the embeddings as well as moving the outlier class back to the 0th position (which is necessary for many other functions). 

Could you test whether it works for you?

**Comment by timdhu** [comment]
Sure! Thanks for making the changes, I will test tomorrow evening
